genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+            throw new DownloadException("Resource not found: " + uriString);
+        }
+
+        final long resourceLastModified = resource.lastModified();
+
+        //Create the resource version dir in cache if it does not exist
+        final File cacheResourceVersionDir = getCacheResourceVersionDir(
+            resourceCacheId,
+            resourceLastModified
+        );
+
+        createDirectoryStructureIfNotExists(cacheResourceVersionDir);
+
+        //Handle to the resource cached locally
+        final File cachedResourceVersionDataFile = getCacheResourceVersionDataFile(
+            resourceCacheId,
+            resourceLastModified
+        );
+
+        try (
+            final FileChannel lockFileChannel = getLockFileChannel(
+                resourceCacheId,
+                resourceLastModified
+            );
+            final Lock lock = lockingFactory.getLock(lockFileChannel);
+        ) {
+            lock.lock();
+            if (!cachedResourceVersionDataFile.exists()) {
+                log.info(
+                    "Cache miss: {} (id: {})",
+                    uriString,
+                    resourceCacheId
+                );
+
+                // Download the resource into the cache
+                final File cachedResourceVersionDownloadFile = getCacheResourceVersionDownloadFile(
+                    resourceCacheId,
+                    resourceLastModified
+                );
+                try (
+                    final InputStream in = resource.getInputStream();
+                    final OutputStream out = new FileOutputStream(cachedResourceVersionDownloadFile)
+                ) {
+                    FileCopyUtils.copy(in, out);
+                    Files.move(cachedResourceVersionDownloadFile, cachedResourceVersionDataFile);
+                }
+            } else {
+                log.info(
+                    "Cache hit: {} (id: {})",
+                    uriString,
+                    resourceCacheId
+                );
+            }
+
+            Files.copy(cachedResourceVersionDataFile, destinationFile);
+        }
+
+        //Clean up any older versions in a different thread
+        oldResourcesCleanUpService.submit(() -> {
+            try {
+                cleanUpOlderResourceVersions(resourceCacheId, resourceLastModified);
+            } catch (IOException e) {
+                log.error(
+                    "Error cleaning up old resource resourceCacheId - {}, resourceLastModified - {}",
+                    resourceCacheId,
+                    resourceLastModified,
+                    e
+                );
+            }
+        });
+    }
+
+    private FileChannel getLockFileChannel(
+        final String resourceCacheId,
+        final long lastModified
+    ) throws FileNotFoundException {
+        return new RandomAccessFile(
+            getCacheResourceVersionLockFile(
+                resourceCacheId,
+                lastModified
+            ),
+            "rws"
```

Comment:
Wow this is garbage class design on the part of `RandomAccessFile`. This should be an enumeration since they only allow certain values anyway but I digress. You should make this string a constant with a good name to clearly identify what it is. I had to go to the javadocs to figure out if it was just some random label or what

Reply:
Ya that part surprised me too. Will do